enhancements/sig-architecture/136-placement-cel-selector/README.md

```diff
@@ -0,0 +1,343 @@
+
+### Goals
+
+- Enhance `Placement` API to allow users to select clusters with CEL expressions.
+
+### Non-Goals
+
+- Enhance the `AddonPlacementScore` to include the original resource value for selecting clusters that is not in the scope.
+
+## Proposal
+
+In this proposal, we will introduce the `celSelector` in the predicate section. It will be a list of `celExpressions`. The result of the expressions are ANDed. The result of `labelSelector`, `claimSelector` and `celSelector` is also ANDed.
+
+We will also introduce a CEL library, `managedCluster` is declared as a variable and can be used in CEL expressions. Custom functions like comparing versions and getting cluster resources will also be added.
+
+### User Stories
+
+1. The user can select clusters by fields other than labels and claims. For example, select clusters by field `managedCluster.Status.version.kubernetes`.
+2. The user can use CEL [Standard macros](https://github.com/google/cel-spec/blob/master/doc/langdef.md#macros) and [Standard functions](https://github.com/google/cel-spec/blob/master/doc/langdef.md#standard-definitions) on the `managedCluster` fields. For example, select clusters with version 1.30.* and 1.31.* by regex `'^1\\.(30|31)\\.\\d+$'`.
+3. The user can use CEL customized function to select clusters by version comparison. For example, select clusters whose version > 1.13.0.
+4. The user can use CEL customized function to select clusters by `AddonPlacementScore`. For example, select clusters whose cpuAvailable < 2.
+
+## Design Details
+
+### API Changes
+
+Add the new field `celSelector` to the `clusterSelector`. `celSelector` will be list of `celExpressions` with string type.
+
+```go
+type ClusterSelector struct {
+	// LabelSelector represents a selector of ManagedClusters by label
+	// +optional
+	LabelSelector metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+	// ClaimSelector represents a selector of ManagedClusters by clusterClaims in status
+	// +optional
+	ClaimSelector ClusterClaimSelector `json:"claimSelector,omitempty"`
+
+	// CelSelector represents a selector of ManagedClusters by CEL expressions on ManagedCluster fields
+	// +optional
+	CelSelector ClusterCelSelector `json:"celSelector,omitempty"`
```

Comment:
Just yesterday a colleague mentioned to me they were adding a string field to a CRD that is expected to contain a CEL expression but that they had no good way to validate if the contents of the string was a valid CEL expression.

Does the same problem exist here?  It looks like it would.

cc @deads2k who has anticipated this need.

cc @cici37

Reply:
The placement API has [condition](https://github.com/open-cluster-management-io/api/blob/main/cluster/v1beta1/types_placement.go#L418 ) in status, so when the CEL evaluation failed, the error will return and report in the [PlacementMisconfigured](https://github.com/open-cluster-management-io/api/blob/main/cluster/v1beta1/types_placement.go#L427C44-L427C66) condition with error message.